File: modlab/ui/main_window.py
from PySide6.QtCore import Qt
from PySide6.QtWidgets import QApplication, QHBoxLayout, QLabel, QMainWindow, QMessageBox, QStackedWidget, QVBoxLayout, QWidget

from ..core.pack import Pack, PackStatus, loader_name
from ..core.pack_manager import PackManager
from ..ipc.backend_process import BackendProcess
from .detail_panel import DetailPanel
from .new_pack_dialog import NewPackDialog
from .pack_grid import PackGrid
from .sidebar import Sidebar
from .top_bar import TopBar


class MainWindow(QMainWindow):
    def __init__(self, python_path: str, script_path: str, parent=None):
        super().__init__(parent)
        self.setWindowTitle("ModLab")
        self.setMinimumSize(1100, 680)
        self.resize(1280, 780)

        self.backend = BackendProcess(python_path, script_path, self)
        self.pack_manager = PackManager(self)
        self.pack_manager.load()
        self._build_layout()
        self._connect_signals()
        self.on_packs_changed()

    def _build_layout(self) -> None:
        central = QWidget(self)
        self.setCentralWidget(central)
        outer_row = QHBoxLayout(central)
        outer_row.setContentsMargins(0, 0, 0, 0)
        outer_row.setSpacing(0)

        self.sidebar = Sidebar(self)
        outer_row.addWidget(self.sidebar)

        main_col = QVBoxLayout()
        main_col.setContentsMargins(0, 0, 0, 0)
        main_col.setSpacing(0)
        self.top_bar = TopBar(self)
        main_col.addWidget(self.top_bar)

        self.stack = QStackedWidget(self)
        library = QWidget()
        lib_row = QHBoxLayout(library)
        lib_row.setContentsMargins(0, 0, 0, 0)
        lib_row.setSpacing(0)
        self.pack_grid = PackGrid(library)
        lib_row.addWidget(self.pack_grid, 1)
        self.detail_panel = DetailPanel(library)
        lib_row.addWidget(self.detail_panel)
        self.stack.addWidget(library)
        self.stack.addWidget(self._make_placeholder_page("Browse Mods", "// mod browser coming soon"))
        self.stack.addWidget(self._make_placeholder_page("Downloads", "// download queue coming soon"))
        self.stack.addWidget(self._make_placeholder_page("Java Manager", "// java runtime manager coming soon"))
        self.stack.addWidget(self._make_placeholder_page("Settings", "// settings coming soon"))
        main_col.addWidget(self.stack, 1)
        outer_row.addLayout(main_col, 1)

    def _make_placeholder_page(self, title: str, sub: str) -> QWidget:
        page = QWidget()
        layout = QVBoxLayout(page)
        layout.setAlignment(Qt.AlignCenter)
        label = QLabel(sub, page)
        label.setObjectName("MutedText")
        label.setStyleSheet("font-size:12px; color:#4a7a96;")
        label.setAlignment(Qt.AlignCenter)
        layout.addWidget(label)
        return page

    def _connect_signals(self) -> None:
        self.sidebar.page_requested.connect(self.on_page_requested)
        self.top_bar.new_pack_requested.connect(self.on_new_pack_requested)
        self.pack_grid.new_pack_requested.connect(self.on_new_pack_requested)
        self.pack_grid.run_requested.connect(self.on_run_requested)
        self.pack_grid.install_requested.connect(self.on_install_requested)
        self.pack_grid.delete_requested.connect(self.on_delete_requested)
        self.pack_grid.pack_selected.connect(self.on_pack_selected)
        self.detail_panel.launch_requested.connect(self.on_run_requested)
        self.detail_panel.add_mods_requested.connect(lambda _key: None)
        self.detail_panel.pack_settings_requested.connect(lambda _key: None)
        self.pack_manager.packs_changed.connect(self.on_packs_changed)
        self.pack_manager.pack_progress_changed.connect(self.on_progress_update)
        self.backend.progress_update.connect(self.on_progress_update)
        self.backend.operation_done.connect(self.on_operation_done)
        self.backend.operation_error.connect(self.on_operation_error)
        self.backend.status_line.connect(self.on_status_line)

    def on_page_requested(self, page: str) -> None:
        page_index = {"library": 0, "browse": 1, "downloads": 2, "java": 3, "settings": 4}
        self.stack.setCurrentIndex(page_index.get(page, 0))
        titles = {
            "library": "Pack Library",
            "browse": "Browse Mods",
            "downloads": "Downloads",
            "java": "Java Manager",
            "settings": "Settings",
        }
        subs = {
            "library": "your instances",
            "browse": "find and install mods",
            "downloads": "active & recent",
            "java": "runtime management",
            "settings": "preferences",
        }
        self.top_bar.set_title(titles.get(page, "ModLab"), subs.get(page, ""))

    def on_new_pack_requested(self) -> None:
        dialog = NewPackDialog(self)
        if dialog.exec() != NewPackDialog.Accepted:
            return
        data = dialog.result_data
        if not data.name or not data.version:
            return
        self.backend.make_pack(data.name, data.version, loader_name(data.loader).lower())

    def on_run_requested(self, key: str) -> None:
        pack = self.pack_manager.find_pack(key)
        if pack is not None and pack.status not in (PackStatus.READY, PackStatus.NEEDS_UPDATE):
            return
        self.backend.run_pack(key)

    def on_install_requested(self, key: str) -> None:
        self.pack_manager.set_status(key, PackStatus.INSTALLING)
        self.backend.install_pack(key)

    def on_delete_requested(self, key: str) -> None:
        confirm = QMessageBox(self)
        confirm.setWindowTitle("Delete Pack")
        confirm.setText(f"Delete <b>{key}</b>? This cannot be undone.")
        confirm.setStandardButtons(QMessageBox.Yes | QMessageBox.Cancel)
        confirm.setDefaultButton(QMessageBox.Cancel)
        app = QApplication.instance()
        if app is not None:
            confirm.setStyleSheet(app.styleSheet())
        if confirm.exec() != QMessageBox.Yes:
            return
        self.backend.delete_pack(key)

    def on_pack_selected(self, key: str) -> None:
        pack = self.pack_manager.find_pack(key)
        if pack is not None:
            self.detail_panel.show_pack(pack)

    def on_packs_changed(self) -> None:
        self.pack_grid.set_packs(self.pack_manager.packs)

    def on_status_line(self, key: str, _message: str) -> None:
        self.pack_manager.set_status(key, PackStatus.INSTALLING)

    def on_progress_update(self, key: str, percent: int) -> None:
        self.pack_manager.set_progress(key, percent)
        pack = self.pack_manager.find_pack(key)
        if pack is not None:
            pack.install_progress = percent
            pack.status = PackStatus.INSTALLING
            self.pack_grid.update_pack(pack)

    def on_operation_done(self, key: str) -> None:
        self.pack_manager.reload_from_disk()
        pack = self.pack_manager.find_pack(key)
        if pack is not None:
            pack.status = PackStatus.READY
            self.pack_grid.update_pack(pack)
            self.detail_panel.show_pack(pack)

    def on_operation_error(self, key: str, message: str) -> None:
        QMessageBox.warning(
            self,
            "ModLab Error",
            f"Operation failed for <b>{key}</b>:<br>{message}",
        )
        self.pack_manager.reload_from_disk()

File: modlab-cpp-client/modlab/src/ui/MainWindow.cpp
#include "MainWindow.h"
#include "NewPackDialog.h"
#include "AeroStyle.h"
#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QLabel>
#include <QMessageBox>
#include <QApplication>

MainWindow::MainWindow(QWidget* parent)
    : QMainWindow(parent)
{
    setWindowTitle("ModLab");
    setMinimumSize(1100, 680);
    resize(1280, 780);

    // ---- Backend process ----
    // python3 by default; override MODLAB_PYTHON env var to use a venv
    QString python = qEnvironmentVariable("MODLAB_PYTHON", "python3");
    QString script = qEnvironmentVariable("MODLAB_SCRIPT", MODLAB_BACKEND_PATH);
    m_backend = new BackendProcess(python, script, this);

    // ---- Pack manager ----
    m_packManager = new PackManager(this);
    m_packManager->load();

    buildLayout();
    connectSignals();

    // Populate grid with whatever is on disk at launch
    onPacksChanged();
}

void MainWindow::buildLayout() {
    auto* central = new QWidget(this);
    setCentralWidget(central);

    auto* outerRow = new QHBoxLayout(central);
    outerRow->setContentsMargins(0,0,0,0);
    outerRow->setSpacing(0);

    // Sidebar
    m_sidebar = new Sidebar(this);
    outerRow->addWidget(m_sidebar);

    // Main column: topbar + stacked pages
    auto* mainCol = new QVBoxLayout;
    mainCol->setContentsMargins(0,0,0,0);
    mainCol->setSpacing(0);

    m_topBar = new TopBar(this);
    mainCol->addWidget(m_topBar);

    // Stacked widget — one widget per page
    m_stack = new QStackedWidget(this);

    // Page 0: Library (grid + detail panel side by side)
    auto* libraryPage = new QWidget;
    auto* libRow = new QHBoxLayout(libraryPage);
    libRow->setContentsMargins(0,0,0,0);
    libRow->setSpacing(0);

    m_packGrid = new PackGrid(libraryPage);
    libRow->addWidget(m_packGrid, 1);

    m_detailPanel = new DetailPanel(libraryPage);
    libRow->addWidget(m_detailPanel);

    m_stack->addWidget(libraryPage);   // index 0

    // Page 1: Browse (placeholder)
    m_browsePage = makePlaceholderPage("Browse Mods", "// mod browser coming soon");
    m_stack->addWidget(m_browsePage);  // index 1

    // Page 2: Downloads (placeholder)
    m_downloadsPage = makePlaceholderPage("Downloads", "// download queue coming soon");
    m_stack->addWidget(m_downloadsPage); // index 2

    // Page 3: Java Manager (placeholder)
    m_javaPage = makePlaceholderPage("Java Manager", "// java runtime manager coming soon");
    m_stack->addWidget(m_javaPage);    // index 3

    // Page 4: Settings (placeholder)
    m_settingsPage = makePlaceholderPage("Settings", "// settings coming soon");
    m_stack->addWidget(m_settingsPage); // index 4

    mainCol->addWidget(m_stack, 1);
    outerRow->addLayout(mainCol, 1);
}

// Small helper to create a centred placeholder for unbuilt pages
QWidget* MainWindow::makePlaceholderPage(const QString& title, const QString& sub) {
    auto* page = new QWidget;
    auto* layout = new QVBoxLayout(page);
    layout->setAlignment(Qt::AlignCenter);

    QLabel* lbl = new QLabel(sub, page);
    lbl->setObjectName("MutedText");
    lbl->setStyleSheet("font-size:12px; color:#4a7a96; letter-spacing:2px;");
    lbl->setAlignment(Qt::AlignCenter);
    layout->addWidget(lbl);
    (void)title;
    return page;
}

void MainWindow::connectSignals() {
    // Sidebar navigation
    connect(m_sidebar, &Sidebar::pageRequested,
            this, &MainWindow::onPageRequested);

    // TopBar
    connect(m_topBar, &TopBar::newPackRequested,
            this, &MainWindow::onNewPackRequested);

    // Pack grid actions
    connect(m_packGrid, &PackGrid::newPackRequested,
            this, &MainWindow::onNewPackRequested);
    connect(m_packGrid, &PackGrid::runRequested,
            this, &MainWindow::onRunRequested);
    connect(m_packGrid, &PackGrid::installRequested,
            this, &MainWindow::onInstallRequested);
    connect(m_packGrid, &PackGrid::deleteRequested,
            this, &MainWindow::onDeleteRequested);
    connect(m_packGrid, &PackGrid::packSelected,
            this, &MainWindow::onPackSelected);

    // Detail panel actions
    connect(m_detailPanel, &DetailPanel::launchRequested,
            this, &MainWindow::onRunRequested);
    connect(m_detailPanel, &DetailPanel::addModsRequested,
            this, [](const QString&){ /* TODO: open mod browser filtered to pack */ });
    connect(m_detailPanel, &DetailPanel::packSettingsRequested,
            this, [](const QString&){ /* TODO: open pack settings dialog */ });

    // PackManager
    connect(m_packManager, &PackManager::packsChanged,
            this, &MainWindow::onPacksChanged);
    connect(m_packManager, &PackManager::packProgressChanged,
            this, [this](const QString& key, int pct){
                onProgressUpdate(key, pct);
            });

    // BackendProcess
    connect(m_backend, &BackendProcess::progressUpdate,
            this, &MainWindow::onProgressUpdate);
    connect(m_backend, &BackendProcess::operationDone,
            this, &MainWindow::onOperationDone);
    connect(m_backend, &BackendProcess::operationError,
            this, &MainWindow::onOperationError);
    connect(m_backend, &BackendProcess::statusLine,
            this, [this](const QString& key, const QString& msg){
                m_packManager->setStatus(key, PackStatus::Installing);
                Q_UNUSED(msg);
            });
}

// ---- Slots ----

void MainWindow::onPageRequested(const QString& page) {
    static const QMap<QString,int> pageIndex = {
        {"library",0}, {"browse",1}, {"downloads",2}, {"java",3}, {"settings",4}
    };
    int idx = pageIndex.value(page, 0);
    m_stack->setCurrentIndex(idx);

    static const QMap<QString,QString> titles = {
        {"library","Pack Library"}, {"browse","Browse Mods"},
        {"downloads","Downloads"},  {"java","Java Manager"},
        {"settings","Settings"}
    };
    static const QMap<QString,QString> subs = {
        {"library","your instances"}, {"browse","find and install mods"},
        {"downloads","active & recent"}, {"java","runtime management"},
        {"settings","preferences"}
    };
    m_topBar->setTitle(titles.value(page,"ModLab"), subs.value(page,""));
}

void MainWindow::onNewPackRequested() {
    NewPackDialog dlg(this);
    if (dlg.exec() != QDialog::Accepted) return;

    NewPackData data = dlg.result();
    if (data.name.isEmpty() || data.version.isEmpty()) return;

    // Immediately mark a placeholder as installing while backend works
    Pack placeholder;
    placeholder.key       = data.name.toLower();
    placeholder.name      = data.name;
    placeholder.mcVersion = data.version;
    placeholder.loader    = data.loader;
    placeholder.status    = PackStatus::Installing;

    // Backend creates + installs
    m_backend->makePack(data.name, data.version, loaderName(data.loader).toLower());

    // Reload from disk once backend signals DONE
    // (handled in onOperationDone)
}

void MainWindow::onRunRequested(const QString& key) {
    if (Pack* p = m_packManager->findPack(key)) {
        if (p->status != PackStatus::Ready && p->status != PackStatus::NeedsUpdate)
            return;
    }
    m_backend->runPack(key);
}

void MainWindow::onInstallRequested(const QString& key) {
    m_packManager->setStatus(key, PackStatus::Installing);
    m_backend->installPack(key);
}

void MainWindow::onDeleteRequested(const QString& key) {
    QMessageBox confirm(this);
    confirm.setWindowTitle("Delete Pack");
    confirm.setText(QString("Delete <b>%1</b>? This cannot be undone.").arg(key));
    confirm.setStandardButtons(QMessageBox::Yes | QMessageBox::Cancel);
    confirm.setDefaultButton(QMessageBox::Cancel);
    confirm.setStyleSheet(qApp->styleSheet());
    if (confirm.exec() != QMessageBox::Yes) return;

    m_backend->deletePack(key);
    // Reload after DONE signal
}

void MainWindow::onPackSelected(const QString& key) {
    if (const Pack* p = m_packManager->findPack(key))
        m_detailPanel->showPack(*p);
}

void MainWindow::onPacksChanged() {
    m_packGrid->setPacks(m_packManager->packs());

    // Keep detail panel in sync if a pack is already showing
    // (it will just refresh with updated data)
}

void MainWindow::onProgressUpdate(const QString& key, int percent) {
    m_packManager->setProgress(key, percent);
    Pack tmp;
    tmp.key             = key;
    tmp.installProgress = percent;
    tmp.status          = PackStatus::Installing;
    if (Pack* p = m_packManager->findPack(key)) {
        p->installProgress = percent;
        m_packGrid->updatePack(*p);
    }
}

void MainWindow::onOperationDone(const QString& key) {
    m_packManager->reloadFromDisk();
    if (Pack* p = m_packManager->findPack(key)) {
        p->status = PackStatus::Ready;
        m_packGrid->updatePack(*p);
        m_detailPanel->showPack(*p);
    }
}

void MainWindow::onOperationError(const QString& key, const QString& msg) {
    QMessageBox::warning(this, "ModLab Error",
        QString("Operation failed for <b>%1</b>:<br>%2").arg(key, msg));
    m_packManager->reloadFromDisk();
}
